### ISSUE TITLE
Defining Aliases Breaks Oh-My-Zsh Features

### ISSUE THREAD

**Issue by akefirad** [DannyBen/alf#35]
It seems the generated code is not quite compatible with Oh-My-Zsh. For example I have my setup so that when I type `kubectl` (or `k` as its alias I had defined in `.zshrc`), I see the current context in the prompt, which is super useful since people have more than one context. But now, when I type `k` (as the alias defined using `alf`) it doesn't show it. Any idea if and how this can be fixed? Thanks.

**Comment by DannyBen** [comment]
Not familiar with Oh-My-Zsh.

I will need reproduction steps to see the problem.
In particular:

1. Sample minimal `~/.zshrc` (if differs from default) or plugin(s) I need to install.
2. Steps to reproduce
3. Expected behavior
4. Actual behavior

Keep in mind that alf creates functions, not aliases. So ZSH (and other shells for that matter) cannot know that `k` is in fact invoking `kubecrl`, and therefore all prompt features that are available for the original command, will not be available in the shortcut.

alf was primarily designed to avoid the need for "tabbing" to see what the command can do.

So bottom line, I highly doubt this can be changed to accommodate what oh-my-zsh is doing.

**Comment by akefirad** [comment]
I see. That's a bummer, because kubestuff was the only reason I was looking for such a tool.
Anyway, in case you wanted to take a look (and maybe find a workaround), [this](https://github.com/akefirad/zsh-in-docker) is a docker image which you can reproduce the issue.
1. Build the image and run it.
2. When you're in, type `k`, you should see `SomeCluster` on the right side of the terminal.
3. If apply alf (`alf save`, the config file is there), and source the aliases, then typing `k` doesn't show the current cluster anymore.  

Please let me know if you need anything else. Also whether you're able to fix this or not. Thanks.

**Comment by DannyBen** [comment]
Cool, and thanks for the docker image - it will be much easier for me to understand and see the issue. I will take a look.

**Comment by DannyBen** [comment]
~Do you know where (which component, path) is that kubectl behavior defined?~

Nevermind, found something here `~/.oh-my-zsh/plugins/kubectl`

**Comment by DannyBen** [comment]
Where is "SomeCluster" coming from?

**Comment by akefirad** [comment]
it comes from `current-context` in `~/.kube/config` file .

**Comment by DannyBen** [comment]
Well, I have no idea how to make it work.

It is not necessarily an `alf` question.

If we try to distill the question to a more generic one, this would be the digest:

How can I specify that a function like the below, should show the same (cluster-) context as `kubectl` does.

```shell
mykube() {
  kubectl # ...
}
```

I will leave this issue open for now, hoping there is some easy answer (for example, like there is for completion: `compdef k=kubectl`).

**Comment by akefirad** [comment]
Should I open an issue on Oh-My-Zsh?

**Comment by DannyBen** [comment]
I am not even sure if its an "Oh My Zsh" of question, or "kubectl plugin" question....

Also, there is some inherent conflict here.
The kubectl zsh plugin comes with tons of aliases on its own.
Alf is designed to "bring your own aliases", and is built in a way that should make it wasy for you to not need the toys and completions of shells like ZSH.

What I am saying is, there are too many components (zsh, oh-my-zsh, kubectl plugin, auto complete, aliases, functions) that compete for the same realestate.

**Comment by akefirad** [comment]
That's right. Probably it's cleaner if I disable kubectl plugin and show the context myself.

**Comment by DannyBen** [comment]
> Probably it's cleaner if I disable kubectl plugin and show the context myself.

I am not sure you have to.
One thing I don't like about that plugin, is all the "root level" aliases they define such as

```shell
alias kdpvc='kubectl describe pvc'
```

This type of nonsense is one of the reasons I made alf in the first place - so that you can do something like:

- `k` instead of `kubectl`
- `d` instead of `describe`

and then it is just `k d pvc` or `k d whatever` or `k describe whatever`

Although I am not a heavy `kubectl` user, I encountered similar issues with `git` and my `g` alias.
Since for example, `g co <tab>` will not show the available branches/tags, as opposed to `git checkout <tab>`. So it is a compromise I am willing to take.

In most cases, it is faster for me to `g co` and type the branch manually, than to type `git checkout <tab>` - but when I have a long branch, I simply do the latter.

One last thing - and in this opinion I am probably different than most - I am not a fan of the fancy prompts and completions. The commands should be short and intuitive enough to type, and if they are not, I am using alf (or [`opcode`](https://github.com/DannyBen/opcode) if it is project specific shortcuts) and I avoid all the complications that come from using ZSH and its plugins.

Whenever I boot a new machine, I install alf, do `alf connect` and I am right at home with all my usual shell.

By the way - if you find yourself installing machines often, another companion tool I am using is [Rush](https://github.com/DannyBen/rush-cli), which is sort of a "personal package manager" that runs shell scripts from GitHub - this is how I am able to have a machine up and running, with not only the shell, but also the software I need, in a very short timeframe (as opposed to regularly searching "how do I install X??" for each package I need. (and here is my own [rush repo](https://github.com/dannyben/rush-repo) with all the packages I need).

**Comment by akefirad** [comment]
Totally agree with you on cryptic aliases generated by the oh-my-zsh plugin. And as I said that's the only reason I tried to find a tool like alf (which was not easy to find BTW 😄).
One hack I found for this specific issue (for the time being) is to enable the prompt by commenting [this](https://github.com/romkatv/powerlevel10k/blob/master/config/p10k-classic.zsh#L1183) line in `~/.p10k.zsh` (it's a P10K issue BTW).

**Comment by akefirad** [comment]
Also thanks for the suggestion. Will check it out.

**Comment by DannyBen** [comment]
Wait - have you tried adding "k" to that line?

```
typeset -g POWERLEVEL9K_KUBECONTEXT_SHOW_ON_COMMAND='k|kubectl|helm|kubens|kubectx|oc|istioctl|kogito|k9s|helmfile'
```

I edited this line in `~/.p10k.zsh`, then sourced the file, and `k` shows 'SomeCluster`, even when alf aliases are sourced.

Can you confirm? Is this a reasonable workaround?

**Comment by akefirad** [comment]
🤔 good point, testing.

**Comment by akefirad** [comment]
Right, ok, seems to be a good workaround. Thanks.

**Comment by DannyBen** [comment]
Excellent! So alf lives on!

BTW - what phrase did you use to search for alf? I am curious to see maybe I can refine some of the texts on the README to make it more findable.

**Comment by akefirad** [comment]
Honestly I don't remember the exact phrases since I tried multiple times to find it. But two things for sure: bash alias maker, bash alias generator (which is included in alf's description 😕). The thing is the result was full of tutorials on how to create aliases, like basic bash tutorials.

**Comment by akefirad** [comment]
One more search phrase I was hoping to get me here was git-style alias maker/generator.
